src/main/kotlin/org/rust/lang/core/psi/ext/RsArrayType.kt

```diff
@@ -18,9 +17,42 @@ val RsArrayType.arraySize: Int? get() {
     return calculateArraySize(expr)
 }
 
-// TODO: support constants and compile time expressions
-fun calculateArraySize(expr: RsExpr?): Int? = (expr as? RsLitExpr)
-    ?.integerLiteral
-    ?.text
-    ?.removeSuffix(TyInteger.Kind.usize.name)
-    ?.toIntOrNull()
+fun calculateArraySize(expr: RsExpr?): Int? = tryEvaluateExpr(expr)
+
+private fun tryEvaluateExpr(expr: RsExpr?): Int? {
+    return when (expr) {
+        is RsLitExpr -> expr.integerLiteral
+            ?.text
+            ?.removeSuffix(TyInteger.Kind.usize.name)
+            ?.toIntOrNull()
+            ?.notNegativeOrNull()
+        is RsPathExpr -> {
+            val const = expr.path.reference.resolve() as? RsConstant ?: return null
```

Comment:
Call of `resolve` method leads to `IndexNotReadyException` because we save array size in stub and try to calculate it while indexing.
I suggest to drop array size filed from `RsArrayTypeStub` because it's more semantic then syntax part of language. And it'll allow to calculate array size expressions, of course:)

Reply:
Hmm, here we need to be very careful. We have an invariant that we looking only at a single file while indexing, so we can't resolve constants (because they can be defined in another files) during it. We need to looking deeply where the indexing need an array size... and is it really need it.

Also we don't want to falling back to AST while resolve (e.g. we resolved a function that return an array with constexpr size, and falled back to AST). Hmm...

And also maybe we want to save the size to stub only when it is trivial (the most case)

Reply:
Also note that `RsConstant` element describes both `const` and `static`!

Reply:
> Hmm, here we need to be very careful. We have an invariant that we looking only at a single file while indexing, so we can't resolve constants during it (because they can be defined in another files).

Yes, I know about this restriction, that's why I've proposed to drop indexing of array size - it becomes useless because of necessity of complex computation.

> Also we don't want to falling back to AST during resolve (e.g. we resolved a function that return an array with constexpr size, and falled back to AST). Hmm...
>
>And also maybe we want to save the size to stub only when it is trivial (the most case)

I think it won't help - we still need to calculate value of integer literal expressions and while it we fall back to AST because we don't create stub for them.
To solve this problem I think we can create stub for literal expressions and save integer value in it. It'll help us not to fall back to AST during type inference but at same time index will contain **all** integers values not only part of array size expr.

Reply:
Although we can save values only for constants and array index expr